File: server-python/app/storage.py
import os
import sqlite3


class Storage:
    def __init__(self, db_path: str) -> None:
        self.db_path = db_path
        self._ensure_database()

    def _connect(self) -> sqlite3.Connection:
        return sqlite3.connect(self.db_path)

    def _ensure_database(self) -> None:
        os.makedirs(os.path.dirname(self.db_path), exist_ok=True)

        with self._connect() as conn:
            cursor = conn.cursor()

            cursor.execute("""
                CREATE TABLE IF NOT EXISTS logins (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    username TEXT NOT NULL,
                    login_timestamp INTEGER NOT NULL
                )
            """)

            cursor.execute("""
                CREATE TABLE IF NOT EXISTS channels (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    name TEXT NOT NULL UNIQUE,
                    created_at INTEGER NOT NULL
                )
            """)

            cursor.execute("""
                CREATE TABLE IF NOT EXISTS messages (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    request_id TEXT,
                    username TEXT NOT NULL,
                    channel_name TEXT NOT NULL,
                    message_text TEXT NOT NULL,
                    sent_timestamp INTEGER NOT NULL
                )
            """)

            self._add_column_if_missing(
                cursor=cursor,
                table_name="messages",
                column_name="request_id",
                column_type="TEXT",
            )

            cursor.execute("""
                CREATE UNIQUE INDEX IF NOT EXISTS idx_messages_request_id
                ON messages(request_id)
            """)

            conn.commit()

    def _add_column_if_missing(
        self,
        cursor,
        table_name: str,
        column_name: str,
        column_type: str,
    ) -> None:
        cursor.execute(f"PRAGMA table_info({table_name})")
        columns = [row[1] for row in cursor.fetchall()]

        if column_name not in columns:
            cursor.execute(
                f"ALTER TABLE {table_name} ADD COLUMN {column_name} {column_type}"
            )

    def save_login(self, username: str, login_timestamp: int) -> None:
        with self._connect() as conn:
            cursor = conn.cursor()
            cursor.execute(
                """
                INSERT INTO logins (username, login_timestamp)
                VALUES (?, ?)
                """,
                (username, login_timestamp),
            )
            conn.commit()

    def channel_exists(self, channel_name: str) -> bool:
        with self._connect() as conn:
            cursor = conn.cursor()
            cursor.execute(
                "SELECT 1 FROM channels WHERE name = ? LIMIT 1",
                (channel_name,),
            )
            return cursor.fetchone() is not None

    def create_channel(self, channel_name: str, created_at: int) -> bool:
        with self._connect() as conn:
            cursor = conn.cursor()
            cursor.execute(
                """
                INSERT OR IGNORE INTO channels (name, created_at)
                VALUES (?, ?)
                """,
                (channel_name, created_at),
            )
            conn.commit()
            return cursor.rowcount > 0

    def list_channels(self) -> list[str]:
        with self._connect() as conn:
            cursor = conn.cursor()
            cursor.execute("SELECT name FROM channels ORDER BY name ASC")
            rows = cursor.fetchall()
            return [row[0] for row in rows]

    def save_message(
        self,
        username: str,
        channel_name: str,
        message_text: str,
        sent_timestamp: int,
        request_id=None,
    ) -> None:
        with self._connect() as conn:
            cursor = conn.cursor()
            cursor.execute(
                """
                INSERT OR IGNORE INTO messages (
                    request_id,
                    username,
                    channel_name,
                    message_text,
                    sent_timestamp
                )
                VALUES (?, ?, ?, ?, ?)
                """,
                (
                    request_id if request_id else None,
                    username,
                    channel_name,
                    message_text,
                    sent_timestamp,
                ),
            )
            conn.commit()

    def list_messages_by_channel(self, channel_name: str) -> list[tuple]:
        with self._connect() as conn:
            cursor = conn.cursor()
            cursor.execute(
                """
                SELECT username, channel_name, message_text, sent_timestamp
                FROM messages
                WHERE channel_name = ?
                ORDER BY sent_timestamp ASC, id ASC
                """,
                (channel_name,),
            )
            return cursor.fetchall()

File: server-java/src/main/java/bbs/Storage.java
package bbs;

import java.io.File;
import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class Storage {
    private final String dbUrl;

    public Storage(String dbPath) {
        File dbFile = new File(dbPath);
        File parent = dbFile.getParentFile();

        if (parent != null) {
            parent.mkdirs();
        }

        this.dbUrl = "jdbc:sqlite:" + dbPath;
        ensureDatabase();
    }

    private Connection connect() throws SQLException {
        return DriverManager.getConnection(dbUrl);
    }

    private void ensureDatabase() {
        try (Connection conn = connect();
             Statement stmt = conn.createStatement()) {

            stmt.execute("""
                CREATE TABLE IF NOT EXISTS logins (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    username TEXT NOT NULL,
                    login_timestamp INTEGER NOT NULL
                )
            """);

            stmt.execute("""
                CREATE TABLE IF NOT EXISTS channels (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    name TEXT NOT NULL UNIQUE,
                    created_at INTEGER NOT NULL
                )
            """);

            stmt.execute("""
                CREATE TABLE IF NOT EXISTS messages (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    request_id TEXT,
                    username TEXT NOT NULL,
                    channel_name TEXT NOT NULL,
                    message_text TEXT NOT NULL,
                    sent_timestamp INTEGER NOT NULL
                )
            """);

            addColumnIfMissing(conn, "messages", "request_id", "TEXT");

            stmt.execute("""
                CREATE UNIQUE INDEX IF NOT EXISTS idx_messages_request_id
                ON messages(request_id)
            """);

        } catch (SQLException e) {
            throw new RuntimeException("Erro ao inicializar banco: " + e.getMessage(), e);
        }
    }

    private void addColumnIfMissing(
            Connection conn,
            String tableName,
            String columnName,
            String columnType
    ) throws SQLException {
        boolean exists = false;

        try (PreparedStatement ps = conn.prepareStatement("PRAGMA table_info(" + tableName + ")");
             ResultSet rs = ps.executeQuery()) {

            while (rs.next()) {
                String existingColumn = rs.getString("name");

                if (existingColumn.equalsIgnoreCase(columnName)) {
                    exists = true;
                    break;
                }
            }
        }

        if (!exists) {
            try (Statement stmt = conn.createStatement()) {
                stmt.execute("ALTER TABLE " + tableName + " ADD COLUMN " + columnName + " " + columnType);
            }
        }
    }

    public void saveLogin(String username, long loginTimestamp) {
        String sql = "INSERT INTO logins (username, login_timestamp) VALUES (?, ?)";

        try (Connection conn = connect();
             PreparedStatement ps = conn.prepareStatement(sql)) {

            ps.setString(1, username);
            ps.setLong(2, loginTimestamp);
            ps.executeUpdate();

        } catch (SQLException e) {
            throw new RuntimeException("Erro ao salvar login: " + e.getMessage(), e);
        }
    }

    public boolean channelExists(String channelName) {
        String sql = "SELECT 1 FROM channels WHERE name = ? LIMIT 1";

        try (Connection conn = connect();
             PreparedStatement ps = conn.prepareStatement(sql)) {

            ps.setString(1, channelName);

            ResultSet rs = ps.executeQuery();

            return rs.next();

        } catch (SQLException e) {
            throw new RuntimeException("Erro ao verificar canal: " + e.getMessage(), e);
        }
    }

    public boolean createChannel(String channelName, long createdAt) {
        String sql = "INSERT OR IGNORE INTO channels (name, created_at) VALUES (?, ?)";

        try (Connection conn = connect();
             PreparedStatement ps = conn.prepareStatement(sql)) {

            ps.setString(1, channelName);
            ps.setLong(2, createdAt);

            int affectedRows = ps.executeUpdate();

            return affectedRows > 0;

        } catch (SQLException e) {
            throw new RuntimeException("Erro ao criar canal: " + e.getMessage(), e);
        }
    }

    public List<String> listChannels() {
        String sql = "SELECT name FROM channels ORDER BY name ASC";
        List<String> channels = new ArrayList<>();

        try (Connection conn = connect();
             PreparedStatement ps = conn.prepareStatement(sql);
             ResultSet rs = ps.executeQuery()) {

            while (rs.next()) {
                channels.add(rs.getString("name"));
            }

            return channels;

        } catch (SQLException e) {
            throw new RuntimeException("Erro ao listar canais: " + e.getMessage(), e);
        }
    }

    public void saveMessage(
            String username,
            String channelName,
            String messageText,
            long sentTimestamp
    ) {
        saveMessage(username, channelName, messageText, sentTimestamp, null);
    }

    public void saveMessage(
            String username,
            String channelName,
            String messageText,
            long sentTimestamp,
            String requestId
    ) {
        String sql = """
            INSERT OR IGNORE INTO messages (
                request_id,
                username,
                channel_name,
                message_text,
                sent_timestamp
            )
            VALUES (?, ?, ?, ?, ?)
        """;

        try (Connection conn = connect();
             PreparedStatement ps = conn.prepareStatement(sql)) {

            if (requestId == null || requestId.isBlank()) {
                ps.setNull(1, Types.VARCHAR);
            } else {
                ps.setString(1, requestId);
            }

            ps.setString(2, username);
            ps.setString(3, channelName);
            ps.setString(4, messageText);
            ps.setLong(5, sentTimestamp);

            ps.executeUpdate();

        } catch (SQLException e) {
            throw new RuntimeException("Erro ao salvar mensagem: " + e.getMessage(), e);
        }
    }
}